tools/chpl-language-server/test/test_server.py

```diff
@@ -0,0 +1,611 @@
+        uri = "file://" + self.tempdir.name
+        added = [WorkspaceFolder(uri=uri, name="Temp Workspace")]
+        event = WorkspaceFoldersChangeEvent(added=added, removed=[])
+        params = DidChangeWorkspaceFoldersParams(event=event)
+        self.client.workspace_did_change_workspace_folders(params)
+
+        return lambda name: TextDocumentIdentifier(
+            uri=f"file://{os.path.join(self.tempdir.name, name + '.chpl')}"
+        )
+
+    def __exit__(self, *exc):
+        return self.tempdir.__exit__(*exc)
+
+
+class SourceFileContext:
+    def __init__(self, main_file: str, client: LanguageClient, **files: str):
+        self.main_file = main_file
+        self.source_files_context = SourceFilesContext(client, **files)
+
+    def __enter__(self):
+        return self.source_files_context.__enter__()(self.main_file)
+
+    def __exit__(self, *exc):
+        return self.source_files_context.__exit__(*exc)
+
+
+def source_files(client: LanguageClient, **files: str):
+    """
+    Context manager that creates a temporary directory and populates
+    it with the given files. The names of the keyword arguments are
+    used as the names of the files. Yields a function that can be used to
+    get the URI of a file by name.
+
+    Also creates a .cls-commands.json file that can be used by the
+    language server to connect the files together in the workspace.
+    """
+    return SourceFilesContext(client, **files)
+
+
+def source_file(client: LanguageClient, contents: str):
+    """
+    Context manager that creates a temporary directory and populates
+    it with the given file. Yields the path to the file.
+    """
+    return SourceFileContext("main", client, main=contents)
+
+
+def pos(coord: typing.Tuple[int, int]):
+    """
+    Shorthand for writing position literals.
+    """
+
+    line, column = coord
+    return Position(line=line, character=column)
+
+
+def standard_module(name: str):
+    """
+    Retrieves the path of a standard module with the given name.
+    """
+
+    return TextDocumentIdentifier(
+        f"file://{os.path.join(CHPL_HOME, 'modules', name)}"
+    )
+
+
+@pytest_lsp.fixture(
+    config=ClientServerConfig(
+        server_command=[
+            sys.executable,
+            os.path.join(
+                CHPL_HOME,
+                "tools",
+                "chpl-language-server",
+                "src",
+                "chpl-language-server.py",
+            ),
+        ]
+    ),
+)
+async def client(lsp_client: LanguageClient):
+    # Setup
+    params = InitializeParams(capabilities=ClientCapabilities())
+    await lsp_client.initialize_session(params)
+
+    yield
+
+    # Teardown
+    await lsp_client.shutdown_session()
+
+
+"""
+The type of 'expected destinations' to be used with the various checking
+functions below.
+"""
+DestinationTestType = typing.Union[
+    None,
+    Position,
+    TextDocumentIdentifier,
+    typing.Tuple[TextDocumentIdentifier, Position],
+]
```

Comment:
I think the docstring comment for the typedef needs to go after the definition, similar to classes/functions